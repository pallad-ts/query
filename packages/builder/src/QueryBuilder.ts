import {Builder} from "@pallad/builder";
import {Either, SchemaValidation, ValidatorError, ViolationsList} from "alpha-validator";
import * as yup from 'yup';
import {createSortableFieldSchema} from "./createSortableFieldSchema";
import {byYup} from "alpha-validator-bridge-yup";
import {QueryConfig} from "./QueryConfig";
import {fromNullable, Maybe} from "@sweet-monads/maybe";

function concatSchema(schema: yup.ObjectSchema<any>, anotherSchema: Maybe<yup.ObjectSchema<any>>) {
	if (anotherSchema.isJust()) {
		return schema.concat(anotherSchema.value);
	}
	return schema;
}

export class QueryBuilder<T extends QueryConfig<any, any>> extends Builder {
	private filtersSchemaObject = yup.object()
		.default({})

	private validation?: (data: unknown) => Promise<Either<ViolationsList, T['query']>>;

	constructor(readonly config: T) {
		super();
	}

	private getPaginationSchema(): Maybe<yup.ObjectSchema<any>> {
		return fromNullable(
			this.config.getPagination()
		)
			.map(x => {
				if (x.type === 'byCursor') {
					return yup.object().shape({
						limit: this.getLimitSchema(x),
						after: yup.string(),
						before: yup.string()
					});
				}
				return yup.object().shape({
					limit: this.getLimitSchema(x),
					offset: yup.number()
						.integer()
						.min(0)
						.default(0)
				});
			});
	}

	private getLimitSchema(paginationConfig: QueryConfig.Pagination) {
		return yup.number()
			.min(1)
			.max(paginationConfig.maxLimit)
			.integer()
			.default(paginationConfig.defaultLimit);
	}

	filtersSchema(schema: yup.ObjectSchema<any>) {
		this.validation = undefined;
		this.filtersSchemaObject = schema.clone().default({});
		return this;
	}

	private getSortingSchema(): Maybe<yup.ObjectSchema<any>> {
		return fromNullable(
			this.config.getSorting()
		).map(x => {
			const sortByType = createSortableFieldSchema(x.sortableFields.slice());
			if (x.type === 'single') {
				return yup.object().shape({
					sortBy: sortByType.default(x.defaultSorting)
				});
			}

			return yup.object().shape({
				sortBy: yup.array()
					.of(sortByType)
					.min(1)
					.default(x.defaultSorting)
			});
		});
	}

	build(input: unknown): Promise<Either<ViolationsList, T['query']>> {
		return this.getValidationFunction()(input);
	}

	async buildOrFail(input: unknown, errorMessage?: string) {
		const result = await this.build(input);
		if (result.isLeft()) {
			throw new ValidatorError(result.value, errorMessage || 'Invalid query');
		}
		return result.value;
	}

	private getValidationFunction() {
		if (!this.validation) {
			let schema = yup.object()
				.shape({
					filters: this.filtersSchemaObject
				})
				.noUnknown();

			schema = concatSchema(schema, this.getPaginationSchema());
			schema = concatSchema(schema, this.getSortingSchema());

			this.validation = SchemaValidation.toValidationFunction<unknown, T['query'], undefined>(
				'query builder',
				byYup<unknown, T['query']>(schema)
			) as (data: unknown) => Promise<Either<ViolationsList, T['query']>>;
		}
		return this.validation;
	}
}

export namespace QueryBuilder {

}
