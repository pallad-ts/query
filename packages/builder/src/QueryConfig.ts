import {PaginableByCursor, PaginableByOffset, Query, SortableFieldDefinition, SortableMulti, SortableSingle} from "@pallad/query";
import {Builder} from "@pallad/builder";
import {QueryBuilder} from "./QueryBuilder";
import {ERRORS} from "./errors";
import {ResultBuilder} from "./ResultBuilder";

function composeDefaultPaginationOptions<T extends Partial<Omit<QueryConfig.Pagination, 'type'>>>(options?: T) {
	return {
		...QueryConfig.defaultPaginationOptions,
		...(options || {})
	};
}

export class QueryConfig<TFilters, TQueryProperties, TResultMeta = unknown> extends Builder {
	private sorting?: QueryConfig.Sorting<any>;
	private pagination?: QueryConfig.Pagination;

	readonly query!: Query<TFilters> & TQueryProperties;
	readonly resultMeta!: TResultMeta;

	static defaultPaginationOptions: Omit<QueryConfig.Pagination, 'type'> = {
		defaultLimit: 50,
		maxLimit: 1000
	};

	cursorPagination(
		options?: Partial<Omit<QueryConfig.Pagination.ByCursor, 'type'>>
	): QueryConfig<TFilters, TQueryProperties & PaginableByCursor, TResultMeta & PaginableByCursor.ResultMeta> {
		this.pagination = {
			type: 'byCursor',
			...composeDefaultPaginationOptions(options)
		};

		return this as any;
	}

	offsetPagination(
		options?: Partial<Omit<QueryConfig.Pagination.ByOffset, 'type'>>
	): QueryConfig<TFilters, TQueryProperties & PaginableByOffset, TResultMeta & PaginableByOffset.ResultMeta> {
		this.pagination = {
			type: 'byOffset',
			...composeDefaultPaginationOptions(options)
		};
		return this as any;
	}

	getPagination() {
		return this.pagination;
	}

	singleSorting<TSortableField extends string>(
		sortableFields: TSortableField[],
		defaultSorting: SortableFieldDefinition<TSortableField>
	) {
		this.sorting = {
			type: 'single',
			sortableFields,
			defaultSorting
		};
		return this as unknown as QueryConfig<TFilters,
			TQueryProperties & SortableSingle<TSortableField>,
			TResultMeta & SortableSingle.ResultMeta<TSortableField>>;
	}

	multiSorting<TSortableField extends string>(
		sortableFields: TSortableField[],
		defaultSorting: Array<SortableFieldDefinition<TSortableField>>
	) {
		this.sorting = {
			type: 'multi',
			sortableFields,
			defaultSorting
		};
		return this as unknown as QueryConfig<TFilters,
			TQueryProperties & SortableMulti<TSortableField>,
			TResultMeta & SortableMulti.ResultMeta<TSortableField>>;
	}

	getSorting() {
		return this.sorting;
	}

	filters<T>() {
		return this as unknown as QueryConfig<T, TQueryProperties>;
	}

	getQueryBuilder(): QueryBuilder<this> {
		this.validate();
		return new QueryBuilder(this);
	}

	validate() {
		if (this.pagination && this.pagination.type === 'byCursor') {
			if (!this.sorting) {
				throw ERRORS.MISSING_SINGLE_SORTING_FOR_CURSOR_PAGINATION();
			}

			if (this.sorting.type === 'multi') {
				throw ERRORS.MULTI_SORTING_FOR_ALLOWED_FOR_CURSOR_PAGINATION();
			}
		}
	}

	getResultBuilder(): ResultBuilder<this> {
		this.validate();
		return new ResultBuilder(this);
	}
}

export namespace QueryConfig {
	export type Pagination = Pagination.ByCursor | Pagination.ByOffset;
	export namespace Pagination {
		export interface ByCursor {
			type: 'byCursor';
			defaultLimit: number;
			maxLimit: number;
		}

		export interface ByOffset {
			type: 'byOffset';
			defaultLimit: number;
			maxLimit: number;
		}
	}

	export type Sorting<TSortableField extends string> = Sorting.Single<TSortableField> | Sorting.Multi<TSortableField>;
	export namespace Sorting {
		export interface Single<TSortableField extends string> {
			type: 'single';
			sortableFields: TSortableField[];
			defaultSorting: SortableFieldDefinition<TSortableField>
		}

		export interface Multi<TSortableField extends string> {
			type: 'multi';
			sortableFields: readonly TSortableField[];
			defaultSorting: Array<SortableFieldDefinition<TSortableField>>;
		}
	}
}
