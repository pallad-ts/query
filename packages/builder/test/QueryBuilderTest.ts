import {QueryConfig} from "@src/QueryConfig";
import {Validation} from "monet";
import * as yup from 'yup';
import {assert, IsExact} from "conditional-type-checks";
import {ViolationsList} from "alpha-validator";
import {PaginableByCursor, PaginableByOffset, Query, SortableMulti, SortableSingle} from "@pallad/query";
import {QueryBuilder} from "@src/QueryBuilder";
import {DEFAULT_SORTING, Filters, SORTABLE_FIELDS} from "./fixtures";

describe('QueryBuilder', () => {
	describe('filters', () => {
		it('types', () => {
			const builder = new QueryConfig()
				.filters<Filters>()
				.getQueryBuilder();

			type Expected = Promise<Validation<ViolationsList, Query<Filters>>>;
			assert<IsExact<ReturnType<typeof builder['build']>, Expected>>(true);
		});

		it('by default accepts all filters', () => {
			const builder = new QueryConfig()
				.getQueryBuilder();

			return expect(builder.build({filters: {literally: 'anything'}}))
				.resolves
				.toEqual(
					Validation.Success({
						filters: {literally: 'anything'}
					})
				);
		});

		describe('uses filters schema if provided', () => {
			const builder = new QueryConfig()
				.filters<Filters>()
				.getQueryBuilder()
				.filtersSchema(yup.object().shape({
					foo: yup.string().oneOf(['foo']).required(),
					bar: yup.string().oneOf(['bar'])
				}));

			it('success', () => {
				return expect(builder.build({
					filters: {
						foo: 'foo',
						bar: 'bar'
					}
				}))
					.resolves
					.toEqual(Validation.Success({
						filters: {foo: 'foo', bar: 'bar'}
					}));
			});

			it('failure', async () => {
				const result = await builder.build({filters: {}});

				expect(result.isFail())
					.toBe(true);

				expect(result.fail())
					.toMatchSnapshot();
			});
		});
	});

	describe('pagination', () => {
		describe.each<[string, QueryBuilder<any>]>([
			[
				'by cursor',
				new QueryConfig()
					.cursorPagination()
					.singleSorting(
						['foo', 'bar'],
						{field: 'foo', direction: 'ASC'}
					)
					.getQueryBuilder()],
			[
				'by offset',
				new QueryConfig()
					.offsetPagination()
					.getQueryBuilder()
			],
		])('common (for %s)', (name, qb) => {
			it('limit cannot be lower than 1', async () => {
				const result = await qb.build({limit: 0})
				expect(result.fail())
					.toMatchSnapshot();
			});

			it('uses default limit', async () => {
				const result = await qb.build({})
				expect(result.success())
					.toMatchObject({limit: 50});
			});

			it('limit cannot be greater than max limit', async () => {
				const result = await qb.build({limit: 5000})
				expect(result.fail())
					.toMatchSnapshot();
			});

			it('limit must be integer', async () => {
				const result = await qb.build({limit: 'asdfa'})
				expect(result.fail())
					.toMatchSnapshot();
			});
		});

		describe('by cursor', () => {
			const builder = new QueryConfig()
				.cursorPagination()
				.singleSorting(SORTABLE_FIELDS.slice(), DEFAULT_SORTING)
				.getQueryBuilder();

			it('types', () => {
				type Expected = Promise<Validation<ViolationsList, Query<unknown> & PaginableByCursor & SortableSingle<'foo' | 'bar'>>>;

				assert<IsExact<ReturnType<typeof builder['build']>, Expected>>(true);
			});

			describe.each([
				['after'],
				['before']
			])('allows "%s" cursor property', property => {
				it('fails if not a string', async () => {
					const result = await builder.build({[property]: []})
					expect(result.fail())
						.toMatchSnapshot();
				});

				it('success', () => {
					const value = `${property} cursor`;
					return expect(builder.build({[property]: value}))
						.resolves
						.toEqual(Validation.Success({
							filters: {},
							[property]: value,
							limit: 50,
							sortBy: DEFAULT_SORTING
						}))
				});
			})
		});

		describe('by offset', () => {
			const builder = new QueryConfig()
				.offsetPagination()
				.getQueryBuilder();

			it('types', () => {
				type Expected = Promise<Validation<ViolationsList, Query<unknown> & PaginableByOffset>>;
				assert<IsExact<ReturnType<typeof builder['build']>, Expected>>(true);
			});

			it('offset by default is 0', () => {
				return expect(builder.build({}))
					.resolves
					.toEqual(
						Validation.Success({
							filters: {},
							limit: 50,
							offset: 0
						})
					);
			});

			it('offset cannot be lower than 0', async () => {
				const result = await builder.build({offset: -1});
				expect(result.fail())
					.toMatchSnapshot();
			});

			it('offset must be an integer', async () => {
				const result = await builder.build({offset: 'asdfasdf'});
				expect(result.fail())
					.toMatchSnapshot();
			});
		});
	});

	describe('sorting', () => {


		describe.each([
			['single', new QueryConfig().singleSorting(SORTABLE_FIELDS.slice(), DEFAULT_SORTING).getQueryBuilder()],
			['multi', new QueryConfig().multiSorting(SORTABLE_FIELDS.slice(), [DEFAULT_SORTING]).getQueryBuilder()]
		])('common (for %s)', (name, qb) => {
			const isMulti = name === 'multi';

			function testCase(sorting: any) {
				return qb.build({sortBy: isMulti ? [sorting] : sorting});
			}

			it('forces to use only allowed sortable fields', async () => {
				const result = await testCase({field: 'unsupported', direction: 'ASC'});

				expect(result.fail())
					.toMatchSnapshot();
			});

			it('sortable field is required', async () => {
				const result = await testCase({direction: 'ASC'});

				expect(result.fail())
					.toMatchSnapshot();
			});

			it('sorting direction is required', async () => {
				const result = await testCase({field: 'foo'});

				expect(result.fail())
					.toMatchSnapshot();
			});

			it('fails for invalid sorting direction', async () => {
				const result = await testCase({field: 'foo', direction: 'bas'});

				expect(result.fail())
					.toMatchSnapshot();
			});


			it.each([
				['ASC'],
				['DESC']
			])('sorting direction must be ASC or DESC', direction => {
				let sorting = {field: 'foo', direction};
				return expect(testCase(sorting))
					.resolves
					.toEqual(Validation.Success({
						filters: {},
						sortBy: isMulti ? [sorting] : sorting
					}));
			})


			it('applies default sorting if not explicitly set', () => {
				return expect(qb.build({}))
					.resolves
					.toEqual(Validation.Success({
						filters: {},
						sortBy: isMulti ? [DEFAULT_SORTING] : DEFAULT_SORTING
					}))
			});
		});

		describe('single', () => {
			const builder = new QueryConfig()
				.singleSorting(SORTABLE_FIELDS.slice(), DEFAULT_SORTING)
				.getQueryBuilder();

			it('types', () => {
				type Expected = Promise<Validation<ViolationsList, Query<unknown> & SortableSingle<'foo' | 'bar'>>>
				assert<IsExact<ReturnType<typeof builder['build']>, Expected>>(true)
			});
		});

		describe('multi', () => {
			const builder = new QueryConfig()
				.multiSorting(SORTABLE_FIELDS.slice(), [DEFAULT_SORTING])
				.getQueryBuilder()

			it('types', () => {
				type Expected = Promise<Validation<ViolationsList, Query<unknown> & SortableMulti<'foo' | 'bar'>>>
				assert<IsExact<ReturnType<typeof builder['build']>, Expected>>(true)
			});

			it('sorting array cannot be empty', async () => {
				const result = await builder.build({
					sortBy: []
				});
				expect(result.fail())
					.toMatchSnapshot();
			});
		});
	});
});

