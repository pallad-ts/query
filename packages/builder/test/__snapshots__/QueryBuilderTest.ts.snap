// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryBuilder filters uses filters schema if provided failure 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "filters.foo is a required field",
      "path": [
        "filters",
        "foo",
      ],
    },
  ],
}
`;

exports[`QueryBuilder pagination by cursor allows "after" cursor property fails if not a string 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "after must be a \`string\` type, but the final value was: \`[]\`.",
      "path": [
        "after",
      ],
    },
  ],
}
`;

exports[`QueryBuilder pagination by cursor allows "before" cursor property fails if not a string 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "before must be a \`string\` type, but the final value was: \`[]\`.",
      "path": [
        "before",
      ],
    },
  ],
}
`;

exports[`QueryBuilder pagination by offset offset cannot be lower than 0 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "offset must be greater than or equal to 0",
      "path": [
        "offset",
      ],
    },
  ],
}
`;

exports[`QueryBuilder pagination by offset offset must be an integer 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "offset must be a \`number\` type, but the final value was: \`NaN\` (cast from the value \`"asdfasdf"\`).",
      "path": [
        "offset",
      ],
    },
  ],
}
`;

exports[`QueryBuilder pagination common (for by cursor) limit cannot be greater than max limit 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "limit must be less than or equal to 1000",
      "path": [
        "limit",
      ],
    },
  ],
}
`;

exports[`QueryBuilder pagination common (for by cursor) limit cannot be lower than 1 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "limit must be greater than or equal to 1",
      "path": [
        "limit",
      ],
    },
  ],
}
`;

exports[`QueryBuilder pagination common (for by cursor) limit must be integer 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "limit must be a \`number\` type, but the final value was: \`NaN\` (cast from the value \`"asdfa"\`).",
      "path": [
        "limit",
      ],
    },
  ],
}
`;

exports[`QueryBuilder pagination common (for by offset) limit cannot be greater than max limit 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "limit must be less than or equal to 1000",
      "path": [
        "limit",
      ],
    },
  ],
}
`;

exports[`QueryBuilder pagination common (for by offset) limit cannot be lower than 1 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "limit must be greater than or equal to 1",
      "path": [
        "limit",
      ],
    },
  ],
}
`;

exports[`QueryBuilder pagination common (for by offset) limit must be integer 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "limit must be a \`number\` type, but the final value was: \`NaN\` (cast from the value \`"asdfa"\`).",
      "path": [
        "limit",
      ],
    },
  ],
}
`;

exports[`QueryBuilder sorting common (for multi) fails for invalid sorting direction 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "sortBy[0].direction must be one of the following values: DESC, ASC",
      "path": [
        "sortBy[0]",
        "direction",
      ],
    },
  ],
}
`;

exports[`QueryBuilder sorting common (for multi) forces to use only allowed sortable fields 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "sortBy[0].field must be one of the following values: foo, bar",
      "path": [
        "sortBy[0]",
        "field",
      ],
    },
  ],
}
`;

exports[`QueryBuilder sorting common (for multi) sortable field is required 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "sortBy[0].field is a required field",
      "path": [
        "sortBy[0]",
        "field",
      ],
    },
  ],
}
`;

exports[`QueryBuilder sorting common (for multi) sorting direction is required 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "sortBy[0].direction is a required field",
      "path": [
        "sortBy[0]",
        "direction",
      ],
    },
  ],
}
`;

exports[`QueryBuilder sorting common (for single) fails for invalid sorting direction 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "sortBy.direction must be one of the following values: DESC, ASC",
      "path": [
        "sortBy",
        "direction",
      ],
    },
  ],
}
`;

exports[`QueryBuilder sorting common (for single) forces to use only allowed sortable fields 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "sortBy.field must be one of the following values: foo, bar",
      "path": [
        "sortBy",
        "field",
      ],
    },
  ],
}
`;

exports[`QueryBuilder sorting common (for single) sortable field is required 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "sortBy.field is a required field",
      "path": [
        "sortBy",
        "field",
      ],
    },
  ],
}
`;

exports[`QueryBuilder sorting common (for single) sorting direction is required 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "sortBy.direction is a required field",
      "path": [
        "sortBy",
        "direction",
      ],
    },
  ],
}
`;

exports[`QueryBuilder sorting multi sorting array cannot be empty 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "sortBy field must have at least 1 items",
      "path": [
        "sortBy",
      ],
    },
  ],
}
`;
